### PR TITLE
fix: improve executeAsync typing to allow callback

### PR DESCRIPTION
**Pull request by DirkoOdendaal**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

Fixing typing for executeAsync to allow callback in typing.
 #6206

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Review by christian-bromann — COMMENTED** [review]
Can we have a test for these typing changes in:

- tests/typings/sync/sync.ts
- tests/typings/webdriverio/async.ts

**Review by mgrybyk — DISMISSED** [review]
can you please also fix sync (in wdio-sync package) version and add a test (tests/typings)?

**Comment by mgrybyk** [comment]
my assumption is you want to make it work like this
```
const result = browser.executeAsync<number>(function (done) {
      done(1)
})
// result is number
```

If so - it doesn't work this way

File: tests/typings/sync/sync.ts [hunk 1]
```diff
@@ -49,6 +49,27 @@ const executeResult = browser.execute(function (x: number) {
 }, 4)
 executeResult.toFixed(2)
 
+const executeAsyncResult = browser.executeAsync(function(a: any, b: any, c: any, d: any, done: any) {
```
Comment (christian-bromann) [inline tests/typings/sync/sync.ts:52]: types should be propagated
```suggestion
const executeAsyncResult = browser.executeAsync(function(a, b, c, d, done) {
```

File: tests/typings/sync/sync.ts [hunk 2]
```diff
@@ -49,6 +49,27 @@ const executeResult = browser.execute(function (x: number) {
...
+    }, 100);
+})
+executeAsyncResultTwo.toEqual(5);
+
+const executeAsyncResultThree = browser.executeAsync(function(done: any) {
+    setTimeout(() => {
+        done('worked')
+    }, 100);
+})
+executeAsyncResultThree.toEqual('worked');
```
Comment (christian-bromann) [inline tests/typings/sync/sync.ts:71]: where does `toEqual` come from?

File: tests/typings/webdriverio/async.ts [hunk 3]
```diff
@@ -89,6 +89,27 @@ async function bar() {
     }, 4)
     executeResult.toFixed(2)
 
+    const executeAsyncResult = await browser.executeAsync(function(a: any, b: any, c: any, d: any, done: any) {
```
Comment (christian-bromann) [inline tests/typings/webdriverio/async.ts:92]: same here, types should be implied by the arguments

File: tests/typings/sync/sync.ts (hunk 2 above)
Comment (christian-bromann) [reply tests/typings/sync/sync.ts:71]: we can just call a primitive function to check that the return type is valid, e.g. for numbers `toFixed` or for strings `charCodeAt`

**Review by christian-bromann — CHANGES_REQUESTED** [review]

**Comment by christian-bromann** [comment]
Also please sign the CLA so we can merge the PR once comments are addressed.

**Review by christian-bromann — APPROVED** [review]
LGTM 👍 

Please sign the CLA so we can move forward with this PR. Thank you!

**Comment by christian-bromann** [comment]
Closed and re-opened PR to see if CLA bot is behind but it seems that you still haven't signed.

File: packages/webdriverio/webdriverio.d.ts [hunk 4]
```diff
@@ -34,16 +34,13 @@ declare namespace WebdriverIO {
             <T>(script: string | ((...arguments: any[]) => T), ...arguments: any[]): Promise<T>;
         };
 
-        // there is no way to add callback as last parameter after `...args`.
```
Comment (christian-bromann) [inline packages/webdriverio/webdriverio.d.ts:37]: Can we also remove this comment about the callback in `packages/wdio-sync/webdriverio.d.ts`?

**Comment by DirkoOdendaal** [comment]
I signed the CLA, so no idea how to progress with this.
https://cla.js.foundation/webdriverio/webdriverio?pullRequest=6207

How else should I sign the CLA?

**Comment by christian-bromann** [comment]
> How else should I sign the CLA?

@DirkoOdendaal make sure that you sign it with the same email you have used to make the commits.

**Comment by DirkoOdendaal** [comment]
> @DirkoOdendaal make sure that you sign it with the same email you have used to make the commits.

@christian-bromann That is what I did. It asks you to login with your github account, so no idea why it would not match.
I've tried again on an incognito window too.

Could you please trigger the check again?

**Comment by christian-bromann** [comment]
@DirkoOdendaal you need to sign the CLA with this email: `entelect@rmal1508122.oc.rabonet.com`

**Comment by DirkoOdendaal** [comment]
> @DirkoOdendaal you need to sign the CLA with this email: `entelect@rmal1508122.oc.rabonet.com`

I might need to recreate the PR. Looks like the username and email was not properly set on my end.
Just triggered the build/tests again to see if my config is setup correctly now.

**Pull request closed without merging**